src/AccountingModule.sol

```diff
@@ -0,0 +1,173 @@
+    error NotStrategy();
+    error InvariantViolation();
+    error TvlTooLow();
+
+    function BASE_ASSET() external view returns (address);
+    function ACCOUNTING_TOKEN() external view returns (AccountingToken);
+
+    function deposit(uint256 amount) external;
+    function withdraw(uint256 amount) external;
+    function processRewards(uint256 amount) external;
+    function processLosses(uint256 amount) external;
+}
+
+contract AccountingModule is IAccountingModule {
+    using SafeERC20 for IERC20;
+
+    uint256 public constant YEAR = 365.25 days;
+    uint256 public constant DIVISOR = 10_000;
+
+    AccountingToken public immutable ACCOUNTING_TOKEN;
+    address public immutable BASE_ASSET;
+    address public immutable STRATEGY;
+
+    address public safe;
+    uint64 public nextRewardWindow;
+    uint16 public cooldownSeconds = 3600;
+    uint16 public targetApy; // in bips;
+    uint16 public lowerBound; // in bips; % of tvl
+
+    modifier checkAndResetCooldown() {
+        if (block.timestamp < nextRewardWindow) revert TooEarly();
+        nextRewardWindow = (uint64(block.timestamp) + cooldownSeconds);
+        _;
+    }
+
+    modifier onlySafeManager() {
+        if (IFlexStrategy(STRATEGY).isSafeManager(msg.sender) == false) revert NotSafeManager();
+        _;
+    }
+
+    modifier onlyStrategy() {
+        if (msg.sender != STRATEGY) revert NotStrategy();
+        _;
+    }
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        address strategy,
+        address baseAsset,
+        address safe_,
+        uint16 targetApy_,
+        uint16 lowerBound_
+    ) {
+        ACCOUNTING_TOKEN = new AccountingToken(name_, symbol_, baseAsset, address(this));
+        BASE_ASSET = baseAsset;
+        STRATEGY = strategy;
+        safe = safe_;
+        targetApy = targetApy_;
+        lowerBound = lowerBound_;
+
+        // TODO: approve on SAFE: IERC20(baseAsset).approve(accountingModule, type(uint256).max);
+    }
+
+    /**
+     * @notice Proxies deposit of base assets from caller to associated SAFE,
+     * and mints an equiv amount of accounting tokens
+     * @param amount amount to deposit
+     */
+    function deposit(uint256 amount) external onlyStrategy {
+        IERC20(BASE_ASSET).safeTransferFrom(msg.sender, safe, amount);
+        ACCOUNTING_TOKEN.mintTo(msg.sender, amount);
+    }
+
+    /**
+     * @notice Proxies withdraw of base assets from associated SAFE to caller,
+     * and burns an equiv amount of accounting tokens
+     * @param amount amount to deposit
+     */
+    function withdraw(uint256 amount) external onlyStrategy {
+        ACCOUNTING_TOKEN.burnFrom(msg.sender, amount);
+        IERC20(BASE_ASSET).safeTransferFrom(safe, msg.sender, amount);
+    }
+
+    /**
+     * @notice Process rewards by minting accounting tokens
+     * @param amount profits to mint
+     */
+    function processRewards(uint256 amount) external onlySafeManager checkAndResetCooldown {
```

Comment:
The onlySafeManager here is overloaded with all responsibilities in the contract

We generally use AccessControlUpgradeable to be able to decouple which address does what in the code.

Many things may be the same but some may vary.

Notice how we decouple all roles here:

https://github.com/yieldnest/yieldnest-vault/blob/eth-max-vault/src/BaseVault.sol#L498

It also helps to have the consistent AccessControlUpgradeable Role Management interface everywhere.

For example, processReports may be called by an automated process or by a coprocessor system.